bson-bench/src/base.js

```diff
@@ -0,0 +1,121 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
+/* eslint-disable @typescript-eslint/no-unused-vars */
+const {
+  Package,
+  BSONLib,
+  BenchmarkSpecification,
+  RunBenchmarkMessage,
+  BenchmarkResult,
+} = require(`${__dirname}/common`);
+/* eslint-enable @typescript-eslint/no-unused-vars */
+
+const { performance } = require("perf_hooks");
+const fs = require("fs");
+const process = require("process");
+const BSON = require("bson");
+
+/**
+ * @param result {BenchmarkResult}
+ * */
+function reportResultAndQuit(result) {
+  if (process.send) process?.send({ type: "returnResult", result });
+  process.disconnect();
+  process.exit(0);
+}
+
+/**
+ * @param error {Error}
+ */
+function reportErrorAndQuit(error) {
+  if (process.send)
+    process.send({
+      type: "returnError",
+      error,
+    });
+  process.disconnect();
+  process.exit(1);
+}
+
+/**
+ * @param bson {BSONLib}
+ * @param config {BenchmarkSpecification}
+ */
+function run(bson, config) {
+  /** @type {((b: Uint8Array, options: any) => any) | ((o: any, options: any) => Uint8Array)} */
+  let fn;
+  /** @type number */
+  let size;
+
+  let doc = BSON.EJSON.parse(fs.readFileSync(config.documentPath, "utf8"));
+
+  switch (config.operation) {
+    case "serialize":
+      fn = bson.serialize;
+      try {
+        size = BSON.calculateObjectSize(doc);
+      } catch (e) {
+        const error = new Error("failed to calculate input object size");
+        error.cause = e;
+        reportErrorAndQuit(error);
+      }
+      break;
+    case "deserialize":
+      fn = bson.deserialize;
+      try {
+        doc = BSON.serialize(doc);
+      } catch (e) {
+        const error = new Error("failed to serialize input object");
+        error.cause = e;
+        reportErrorAndQuit(error);
+      }
+      break;
+    default:
+      reportErrorAndQuit(new Error("unknown test type"));
+  }
+
+  // Check if we can successfully run function under test and report failure if not
+  try {
+    fn(doc, config.options);
+  } catch (e) {
+    const error = new Error("operation under test failed");
+    error.cause = e;
```

Comment:
Can do `new Error ("message", {cause})` since our minimum supported version is 16.20.1 and that was added in 16.9